src/state_machine.rs

```diff
@@ -0,0 +1,1062 @@
+        push(value = 0),
+        push(value = -1),
+        pop(()),
+        pop(())
+    ],
+    machine: "<elided for clarity>",
+    trace: false
+}
+        successes: 0
+        local rejects: 0
+        global rejects: 0
+', examples/state_machine_v1.rs:174:5
+```
+
+Here, we can see our initial state (always an empty heap) and more importantly,
+what sequence of mutations we performed. In the minimal failing case, we pushed
+the integer `0` twice, then `-1` once, then tried to pop two values. The second
+pop however returned `-1` while `0` was still in the heap, violating the
+contract for a max heap.
+
+To get better visibility into what's going on during the test cases, we can
+change our test body to
+
+```rust,ignore
+        test_case.trace().run()?;
+```
+
+which results in the below additional output. Here, we can not only see what
+mutations occurred, but also what the state of the system was at each step.
+
+```text
+initial state: MyHeapState { heap: MyHeap { data: [] }, existing_elements: [] }
+check initial invariants
+enter push(value = 0)
+        => MyHeapState { heap: MyHeap { data: [0] }, existing_elements: [0] }
+exit  push(value = 0)
+enter push(value = 0)
+        => MyHeapState { heap: MyHeap { data: [0, 0] }, existing_elements: [0, 0] }
+exit  push(value = 0)
+enter push(value = -1)
+        => MyHeapState { heap: MyHeap { data: [0, 0, -1] }, existing_elements: [0, 0, -1] }
+exit  push(value = -1)
+enter pop(())
+        => MyHeapState { heap: MyHeap { data: [-1, 0] }, existing_elements: [-1, 0] }
+exit  pop(())
+enter pop(())
+thread 'main' panicked at 'Popped value -1, which was less than 0 still in the heap', examples/state_machine_v1.rs:149:25
+```
+
+*/
+#[macro_export]
+macro_rules! prop_state_machine {
+    (type = $statety:ty;
+     $($token:tt)*) => {
+        #[allow(unused_mut)]
+        let mut ops = $crate::state_machine::helper::new_ops();
+        #[allow(unused_mut)]
+        let mut invariants = $crate::state_machine::helper::new_invariant_vec();
+
+        prop_sm_helper! {
+            @_MUNCH(ops, invariants, $statety) $($token)*
+        }
+
+        $crate::state_machine::StateMachine::new_from_vecs(
+            ops, invariants)
+    };
+}
+
+#[macro_export]
+#[allow(missing_docs)]
+#[doc(hidden)]
+macro_rules! prop_sm_helper {
+    (@_MUNCH($ops:ident, $invariants:ident, $statety:ty)) => { };
+
+    (@_MUNCH($ops:ident, $invariants:ident, $statety:ty)
+     $(#[$attr:ident = $attrv:expr])*
+     fn $op_name:ident(&mut $statevar:ident) $body:block
+     $($rest:tt)*) => {
+        #[allow(unused_mut)]
+        let mut attrs = $crate::state_machine::helper::OpAttrs::DEFAULT;
+        $(attrs.$attr = $attrv;)*
+        $ops.push(
+            (attrs.weight,
+             $crate::strategy::Strategy::boxed(
+                 $crate::strategy::LazyJust::new(
+                     || $crate::state_machine::Mutator::boxed(
+                         $crate::state_machine::DirectMutator::new(
```

Comment:
These series of calls could perhaps add some pressure on compile times;
Consider creating a hidden exported helper function for this?